### PR TITLE
Migrate to lucene 8.0.0

### PR DESCRIPTION
**Pull request by mkleen**
## Checklist

 - [x] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [x] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed

### CODE REVIEW

File: es/es-analysis-common/src/main/java/org/elasticsearch/analysis/common/CommonAnalysisPlugin.java [hunk 1]
```diff
@@ -391,15 +390,7 @@
...
         filters.add(PreConfiguredTokenFilter.singleton("dutch_stem", false, input -> new SnowballFilter(input, new DutchStemmer())));
         filters.add(PreConfiguredTokenFilter.singleton("edge_ngram", false, input ->
-                new EdgeNGramTokenFilter(input, EdgeNGramTokenFilter.DEFAULT_MIN_GRAM_SIZE, EdgeNGramTokenFilter.DEFAULT_MAX_GRAM_SIZE)));
-        filters.add(PreConfiguredTokenFilter.singletonWithVersion("edgeNGram", false, (reader, version) -> {
-            if (version.onOrAfter(org.elasticsearch.Version.ES_V_6_5_1)) {
-                DEPRECATION_LOGGER.deprecatedAndMaybeLog("edgeNGram_deprecation",
-                        "The [edgeNGram] token filter name is deprecated and will be removed in a future version. "
-                                + "Please change the filter name to [edge_ngram] instead.");
-            }
-            return new EdgeNGramTokenFilter(reader, EdgeNGramTokenFilter.DEFAULT_MIN_GRAM_SIZE, EdgeNGramTokenFilter.DEFAULT_MAX_GRAM_SIZE);
```
Comment (seut) [inline es/es-analysis-common/src/main/java/org/elasticsearch/analysis/common/CommonAnalysisPlugin.java]: we would have to deprecate it first before we can remove it. also `edgeNGram` is still used at `getTokenFilters()`

File: es/es-analysis-common/src/main/java/org/elasticsearch/analysis/common/CommonAnalysisPlugin.java [hunk 2]
```diff
@@ -415,15 +406,7 @@
...
                         LimitTokenCountFilterFactory.DEFAULT_CONSUME_ALL_TOKENS)));
-        filters.add(PreConfiguredTokenFilter.singleton("ngram", false, NGramTokenFilter::new));
-        filters.add(PreConfiguredTokenFilter.singletonWithVersion("nGram", false, (reader, version) -> {
-            if (version.onOrAfter(org.elasticsearch.Version.ES_V_6_5_1)) {
-                DEPRECATION_LOGGER.deprecatedAndMaybeLog("nGram_deprecation",
-                        "The [nGram] token filter name is deprecated and will be removed in a future version. "
-                                + "Please change the filter name to [ngram] instead.");
-            }
-            return new NGramTokenFilter(reader);
-        }));
```
Comment (seut) [inline es/es-analysis-common/src/main/java/org/elasticsearch/analysis/common/CommonAnalysisPlugin.java]: same here, not deprecated and still in use

File: es/es-analysis-common/src/main/java/org/elasticsearch/analysis/common/CommonAnalysisPlugin.java [hunk 3]
```diff
@@ -415,15 +406,7 @@
...
+//        filters.add(PreConfiguredTokenFilter.singleton("ngram", false, NGramTokenFilterFactory::new));
```
Comment (seut) [inline es/es-analysis-common/src/main/java/org/elasticsearch/analysis/common/CommonAnalysisPlugin.java]: wonder why this does not result in test failures

File: es/es-analysis-common/src/main/java/org/elasticsearch/analysis/common/CommonAnalysisPlugin.java [hunk 4]
```diff
@@ -482,7 +465,7 @@
             () -> new EdgeNGramTokenizer(EdgeNGramTokenizer.DEFAULT_MIN_GRAM_SIZE, EdgeNGramTokenizer.DEFAULT_MAX_GRAM_SIZE), null));
         tokenizers.add(PreConfiguredTokenizer.singleton("pattern", () -> new PatternTokenizer(Regex.compile("\\W+", null), -1), null));
         tokenizers.add(PreConfiguredTokenizer.singleton("thai", ThaiTokenizer::new, null));
-        tokenizers.add(PreConfiguredTokenizer.singleton("lowercase", LowerCaseTokenizer::new, () -> new TokenFilterFactory() {
+        tokenizers.add(PreConfiguredTokenizer.singleton("lowercase", XLowerCaseTokenizer::new, () -> new TokenFilterFactory() {
```
Comment (seut) [inline es/es-analysis-common/src/main/java/org/elasticsearch/analysis/common/CommonAnalysisPlugin.java]: Usage of a LowerCase tokenizer is deprecated, a LowerCase filter should be used instead. We should document this deprecation inside at our CHANGES.

File: es/es-analysis-common/src/main/java/org/elasticsearch/analysis/common/LowerCaseTokenizerFactory.java [hunk 5]
```diff
@@ -20,13 +20,15 @@
...
-import org.apache.lucene.analysis.core.LowerCaseTokenizer;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.env.Environment;
 import org.elasticsearch.index.IndexSettings;
 import org.elasticsearch.index.analysis.AbstractTokenizerFactory;
 import org.elasticsearch.index.analysis.MultiTermAwareComponent;
 
+@Deprecated
+// Elastic set it to deprecated and will remove it, LowerCaseTokenizer is removed in Lucene 8.0.0,
+// a LowerCaseFilter should be used instead
```
Comment (seut) [inline es/es-analysis-common/src/main/java/org/elasticsearch/analysis/common/LowerCaseTokenizerFactory.java]: Can we use proper javadoc syntax here pls? Also I think doc should be above any annotation.

File: es/es-analysis-common/src/main/java/org/elasticsearch/analysis/common/WordDelimiterGraphTokenFilterFactory.java [hunk 6]
```diff
@@ -45,12 +49,16 @@
...
+        new DeprecationLogger(LogManager.getLogger(WordDelimiterGraphTokenFilterFactory.class));
+
     private final byte[] charTypeTable;
     private final int flags;
     private final CharArraySet protoWords;
+    private final boolean adjustOffsets;
 
     public WordDelimiterGraphTokenFilterFactory(IndexSettings indexSettings, Environment env,
-            String name, Settings settings) {
+                                                String name, Settings settings) {
```
Comment (seut) [inline es/es-analysis-common/src/main/java/org/elasticsearch/analysis/common/WordDelimiterGraphTokenFilterFactory.java]: Either all in one line, or one argument per line pls.

File: es/es-analysis-common/src/main/java/org/elasticsearch/analysis/common/WordDelimiterGraphTokenFilterFactory.java [hunk 7]
```diff
@@ -88,11 +96,24 @@ public WordDelimiterGraphTokenFilterFactory(IndexSettings indexSettings, Environ
...
+    }
+
+    @Override
+    public TokenFilterFactory getSynonymFilter() {
+        if (indexSettings.getIndexVersionCreated().onOrAfter(Version.V_4_0_0)) {
+            throw new IllegalArgumentException("Token filter [" + name() + "] cannot be used to parse synonyms");
+        }
+        else {
+            DEPRECATION_LOGGER.deprecatedAndMaybeLog("synonym_tokenfilters", "Token filter [" + name()
+                                                                             + "] will not be usable to parse synonyms after v7.0");
```
Comment (seut) [inline es/es-analysis-common/src/main/java/org/elasticsearch/analysis/common/WordDelimiterGraphTokenFilterFactory.java]: what does `after v7.0` mean? We should highlight deprecations at the CHANGES.

File: es/es-server/src/main/java/org/apache/lucene/queries/BlendedTermQuery.java [hunk 8]
```diff
@@ -304,8 +308,8 @@ protected Query topLevelQuery(Term[] terms, TermContext[] ctx, int[] docFreqs, i
                         query = new BoostQuery(query, boosts[i]);
                     }
                     if ((maxTermFrequency >= 1f && docFreqs[i] > maxTermFrequency)
-                            || (docFreqs[i] > (int) Math.ceil(maxTermFrequency
-                            * maxDoc))) {
+                        || (docFreqs[i] > (int) Math.ceil(maxTermFrequency
+                                                          * maxDoc))) {
```
Comment (seut) [inline es/es-server/src/main/java/org/apache/lucene/queries/BlendedTermQuery.java]: pls move up, no need for putting it in a new line here.

File: es/es-server/src/main/java/org/elasticsearch/common/lucene/search/Queries.java [hunk 9]
```diff
@@ -56,9 +61,17 @@ public static Query newLenientFieldQuery(String field, RuntimeException e) {
...
-        return new DocValuesFieldExistsQuery(SeqNoFieldMapper.PRIMARY_TERM_NAME);
+    public static Query newNonNestedFilter(Version indexVersionCreated) {
+        if (indexVersionCreated.onOrAfter(Version.ES_V_6_1_4)) {
+            return new DocValuesFieldExistsQuery(SeqNoFieldMapper.PRIMARY_TERM_NAME);
+        } else {
+            return new BooleanQuery.Builder()
+                .add(new MatchAllDocsQuery(), Occur.FILTER)
+//                .add(newNestedFilter(), Occur.MUST_NOT)
+                .build();
+        }
```
Comment (seut) [inline es/es-server/src/main/java/org/elasticsearch/common/lucene/search/Queries.java]: why is this needed? looks incomplete and can be maybe skipped/removed as we do not support nested fields.

File: es/es-server/src/main/java/org/elasticsearch/common/lucene/search/Queries.java [hunk 10]
```diff
@@ -135,7 +148,7 @@ public static int calculateMinShouldMatch(int optionalClauseCount, String spec)
                     return result;
                 } else {
                     result = calculateMinShouldMatch
-                            (optionalClauseCount, parts[1]);
+                        (optionalClauseCount, parts[1]);
```
Comment (seut) [inline es/es-server/src/main/java/org/elasticsearch/common/lucene/search/Queries.java]: pls move up, strange style ;)

File: es/es-server/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java [hunk 11]
```diff
@@ -141,8 +143,8 @@ boolean recoverFromLocalShards(BiConsumer<String, MappingMetaData> mappingUpdate
     }
 
     void addIndices(final RecoveryState.Index indexRecoveryStats, final Directory target, final Sort indexSort, final Directory[] sources,
-            final long maxSeqNo, final long maxUnsafeAutoIdTimestamp, IndexMetaData indexMetaData, int shardId, boolean split) throws IOException {
-
+                    final long maxSeqNo, final long maxUnsafeAutoIdTimestamp, IndexMetaData indexMetaData, int shardId, boolean split,
+                    boolean hasNested) throws IOException {
```
Comment (seut) [inline es/es-server/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java]: can we format this properly by putting each arg in a line?

File: es/es-testing/src/main/java/org/elasticsearch/indices/analysis/AnalysisFactoryTestCase.java [hunk 12]
```diff
@@ -296,160 +296,6 @@ public AnalysisFactoryTestCase(AnalysisPlugin plugin) {
...
-        Set<Object> classesMissingMultiTermSupport = new HashSet<>(expected);
-        classesMissingMultiTermSupport.removeAll(actual);
-        assertTrue("Pre-built components are missing multi-term support: " + classesMissingMultiTermSupport,
-                classesMissingMultiTermSupport.isEmpty());
-
-        Set<Object> classesThatShouldNotHaveMultiTermSupport = new HashSet<>(actual);
-        classesThatShouldNotHaveMultiTermSupport.removeAll(expected);
-        assertTrue("Pre-built components should not have multi-term support: " + classesThatShouldNotHaveMultiTermSupport,
-                classesThatShouldNotHaveMultiTermSupport.isEmpty());
-    }
```
Comment (seut) [inline es/es-testing/src/main/java/org/elasticsearch/indices/analysis/AnalysisFactoryTestCase.java]: why is this removed only? afaik the whole class is not used currently

File: sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneBatchIterator.java [hunk 13]
```diff
@@ -79,7 +80,7 @@
     public LuceneBatchIterator(IndexSearcher indexSearcher,
                                Query query,
                                @Nullable Float minScore,
-                               boolean doScores,
+                                   boolean doScores,
```
Comment (seut) [inline sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneBatchIterator.java]: wrong indent

File: sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java [hunk 14]
```diff
@@ -132,10 +132,7 @@ public void close() {
         TopFieldCollector topFieldCollector = TopFieldCollector.create(
             sort,
             batchSize,
-            true,
-            doDocsScores,
-            doDocsScores,
-            false           // trackTotalHits - we don't use the number of total hits
+            batchSize
```
Comment (seut) [inline sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java]: as we're not interested in the total hits, we should pass `1` here to not slow down the query. also a comment similar to the removed one would be great.

File: sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java [hunk 15]
```diff
@@ -152,11 +149,9 @@ public void close() {
             sort,
             batchSize,
             lastDoc,
-            true,
-            doDocsScores,
-            doDocsScores,
-            false           // trackTotalHits - we don't use the number of total hits
+            batchSize
```
Comment (seut) [inline sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java]: same here, `1` should be used instead (and comment pls)

File: sql/src/test/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollectorTest.java [hunk 16]
```diff
@@ -286,6 +287,7 @@ public void testSearchMoreAppliesMinScoreFilter() throws Exception {
         addDoc(w, fieldType, "Arthurr"); // not "Arthur" to lower score
         w.commit();
         IndexSearcher searcher = new IndexSearcher(DirectoryReader.open(w, true, true));
+        searcher.setSimilarity(new LegacyBM25Similarity());
```
Comment (seut) [inline sql/src/test/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollectorTest.java]: can't we adapt the test instead of using an old/deprecated similarity algo?

File: sql/src/test/java/io/crate/integrationtests/UnionIntegrationTest.java [hunk 17]
```diff
@@ -23,6 +23,7 @@
 package io.crate.integrationtests;
 
 import io.crate.testing.TestingHelpers;
+import io.crate.testing.UseJdbc;
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/UnionIntegrationTest.java]: unused import

File: es/es-analysis-common/src/main/java/org/elasticsearch/analysis/common/CommonAnalysisPlugin.java (hunk 1 above)
Comment (mkleen) [reply es/es-analysis-common/src/main/java/org/elasticsearch/analysis/common/CommonAnalysisPlugin.java]: Its neither deprecated nor removed, but i found the new defaults for the configuration.

File: blackbox/docs/appendices/release-notes/unreleased.rst [hunk 18]
```diff
@@ -41,6 +41,10 @@ Unreleased Changes
 Breaking Changes
 ================
 
+- LowerCase tokenizer will be removed
```
Comment (seut) [inline blackbox/docs/appendices/release-notes/unreleased.rst]: We are using the past tense here. Will it be removed? I thought its deprecated.

File: sql/src/test/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollectorTest.java [hunk 19]
```diff
@@ -299,19 +297,19 @@ public void testSearchMoreAppliesMinScoreFilter() throws Exception {
...
+            assertThat((float) row.get(0), Matchers.greaterThanOrEqualTo(0.15f));
             count++;
         }
         assertThat(count, is(2));
 
         count = 0;
         // searchMore -> 1 row is below minScore
         for (Row row : collector.collect()) {
-            assertThat((float) row.get(0), Matchers.greaterThanOrEqualTo(0.30f));
+            assertThat((float) row.get(0), Matchers.greaterThanOrEqualTo(0.15   f));
```
Comment (seut) [inline sql/src/test/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollectorTest.java]: minor: spaces should be removed

File: sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java [hunk 20]
```diff
@@ -168,7 +160,9 @@ public void close() {
             collector = new MinimumScoreCollector(collector, minScore);
         }
         searcher.search(query, collector);
-        return scoreDocToIterable(topFieldCollector.topDocs().scoreDocs);
+        ScoreDoc[] scoreDocs = topFieldCollector.topDocs().scoreDocs;
+        TopFieldCollector.populateScores(scoreDocs, searcher, query);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java]: Does this mean we're now always scoring even if the former `doDocsScores` flag would be false?

Comment (mkleen) [reply sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java]: Well done, thanks for catching this. This was my mistake, since the boolean trackDocScores was removed from TopFieldCollector#create and then additionally TopFieldCollector#populateScores introduced.

**Comment by mkleen** [comment]
> Should we have an entry about the lucene upgrade in the release notes? At least the scoring change is something that could affect users.

Yes, we should do this.

File: sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java (hunk 20 above)
Comment (mkleen) [reply sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java]: Since the refactoring was done in multiple steps i first removed it, since TopFieldCollector#create wasn't using it anymore and then removed it also from LuceneOrderedDocCollector.

Comment (mkleen) [reply sql/src/main/java/io/crate/execution/engine/collect/collectors/LuceneOrderedDocCollector.java]: I fixed it, and added a test to LuceneOrderedDocCollectorTest to cover the doDocsScores flag now.

File: blackbox/docs/appendices/release-notes/unreleased.rst [hunk 21]
```diff
@@ -137,6 +141,8 @@ Deprecations
 Changes
 =======
 
+- Upgrade to Lucene 8.0.0
```
Comment (mfussenegger) [inline blackbox/docs/appendices/release-notes/unreleased.rst]: Can we elaborate on the user facing consequences (scoring change?), are there any others?

Comment (mkleen) [reply blackbox/docs/appendices/release-notes/unreleased.rst]: Not that I am aware of. We should decide on the codec format what will be supported in terms of backward compatibility.

Comment (mfussenegger) [inline blackbox/docs/appendices/release-notes/unreleased.rst]:
```suggestion
- Upgraded Lucene to 8.0.0. This includes a change in the scoring logic, so fulltext queries including ``_score`` filters may behave slightly different. 
```

**Comment by mkleen** [comment]
retest this please

**Comment by mkleen** [comment]
retest this please

**Comment by mkleen** [comment]
retest this please

**Comment by mkleen** [comment]
retest this please

**Comment by mkleen** [comment]
retest this please
